### ISSUE TITLE
After latest update, zsh completion gets corrupted

### ISSUE THREAD

**Issue by jaraco** [Homebrew/brew#1027]
- [X] Ran `brew update` and retried your prior step?
- [X] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [ ] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?

### Bug reports:

After updating homebrew this morning, I ran into an issue similar to the one described in HomeBrew/legacy-homebrew#45811. Except I'm not trying to do anything with root. I'm just loading my shell. When I start the shell, it just emits the functions.

I delete .zcompdump and start a new shell and I don't get the output, but shell completion isn't working either. .zcompdump gets regenerated, and on the next shell, the functions are emitted again.

Running `compinit -i` also emits the same. Here's what my .zshrc looks like:

```
$ cat .zshrc
# Lines configured by zsh-newuser-install
HISTFILE=~/.histfile
HISTSIZE=1000
SAVEHIST=1000
setopt appendhistory autocd beep extendedglob nomatch notify
bindkey -e
# End of lines configured by zsh-newuser-install
# The following lines were added by compinstall
zstyle :compinstall filename '/Users/jaraco/.zshrc'

autoload -Uz compinit
compinit -i
# End of lines added by compinstall

. other/stuff
```

The reason I'm using `compinit -i` is because:

```
$ compaudit
There are insecure directories:
/usr/local/share/zsh/site-functions
/usr/local/Cellar/zsh/5.2/share/zsh/functions
/usr/local/share/zsh
/usr/local/Cellar/zsh/5.2/share/zsh
```

Those directories aren't really insecure, they're just writable by the group admin:

```
$ ls -lad /usr/local/share/zsh/site-functions
drwxrwxr-x  8 root  admin  272 Sep 19 08:54 /usr/local/share/zsh/site-functions
```

I'll continue to investigate, but I'd appreciate advice.

**Comment by jaraco** [comment]
I should also mention that in addition to updating Homebrew, I upgrade to macOS Sierra (10.12) yesterday. That upgrade caused the group write bit to be cleared on /usr/local so I did also this morning `chmod -R g+w /usr/local` to restore homebrew access in order to do the update.

**Comment by MikeMcQuaid** [comment]
> Those directories aren't really insecure, they're just writable by the group admin:

ZSH considers that insecure. Nothing else we can do about that, sorry 😭

**Comment by zmwangx** [comment]
Security check can be skipped. Read http://zsh.sourceforge.net/Doc/Release/Completion-System.html (or just search for "security" in it).

**Comment by jaraco** [comment]
I tried satisfying compaudit, but that doesn't help:

```
$ sudo chmod g-w /usr/local/share/zsh/site-functions
$ sudo chmod g-w /usr/local/Cellar/zsh/5.2/share/zsh/functions
$ sudo chmod g-w /usr/local/share/zsh                         
$ sudo chmod g-w /usr/local/Cellar/zsh/5.2/share/zsh          
$ compaudit
$ compinit
compaudit () {
	emulate -L zsh
	setopt extendedglob
...
```

**Comment by jaraco** [comment]
Aha! After changing those settings, removing the corrupt .zcompdump, and restarting the shell, I have completion again.

**Comment by zmwangx** [comment]
Please show me the output of

```sh
whoami && /bin/ls -ld /usr/local/share/zsh/site-functions /usr/local/Cellar/zsh/5.2/share/zsh/functions /usr/local/share/zsh /usr/local/Cellar/zsh/5.2/share/zsh
```

**Comment by zmwangx** [comment]
Okay, glad to know.

**Comment by jaraco** [comment]
Ugh. After changing the permissions on those directories to make them group read-only (as above), Homebrew started raising errors like:

```
Error: Failed to link all completions, docs and manpages:
  Permission denied - /usr/local/share/zsh/site-functions/_brew
```

And indeed, `/usr/local/share/zsh/site-functions` is owned by `root:admin` and to satisfy compaudit, not group writable. I'd really rather not _force_ the issue by simply claiming user ownership of that directory - after all, it's not in my home directory, and could easily be used by another user logged into the system.

I'd like to be able to rely on the group membership to grant write permission.

I did find that I could restore group write to that directory and the errors in Homebrew went away.

```
$ sudo chmod g+w /usr/local/share/zsh/site-functions
$ sudo chmod g+w /usr/local/share/zsh
```

I also double-checked and the issue with compinit is not recurring.

So I also restored group write to the Cellar folders, but doing so brings back the corruption to `.zcompdump` and the failure in subsequent shell launches.

```
$ sudo chmod g+w /usr/local/Cellar/zsh/5.2/share/zsh  
$ sudo chmod g+w /usr/local/Cellar/zsh/5.2/share/zsh/functions
$ zsh
$ zsh
compdump () {
	# undefined
	builtin autoload -XUz
}
...
```

So it seems I can continue to work around the issue by leaving the /usr/local/share directories group writable and the /usr/local/Cellar directories not group writable. Interestingly, as the Cellar directories are owned by me, they won't need to be group writable to function:

```
$ sudo chmod g-w /usr/local/Cellar/zsh/5.2/share/zsh         
$ sudo chmod g-w /usr/local/Cellar/zsh/5.2/share/zsh/functions
$ ls -lad /usr/local/Cellar/zsh/5.2/share/zsh /usr/local/Cellar/zsh/5.2/share/zsh/functions
drwxr-xr-x    5 jaraco  admin    170 Dec  3  2015 /usr/local/Cellar/zsh/5.2/share/zsh
drwxr-xr-x  982 jaraco  admin  33388 Apr 15 05:46 /usr/local/Cellar/zsh/5.2/share/zsh/functions
$ whoami
jaraco
```

To summarize, I'm able to work around the issue with what feels like a delicate balance of adjustments to the permissions, and I still don't know what's causing the underlying corruption to .zcompdump, though it does seem to be related to group write permissions on `/usr/local/Cellar/...`.

**Comment by jaraco** [comment]
> Security check can be skipped. Read http://zsh.sourceforge.net/Doc/Release/Completion-System.html (or just search for "security" in it).

For clarity, let me re-state that my .zshrc invokes compinit with `-i`, so already should be bypassing the security checks.

**Comment by apjanke** [comment]
As a workaround, could you use ACLs (in the Finder info window under "Sharing & Permissions") to grant yourself (the user that runs `brew`) write permissions to `/usr/local/share/zsh/site-functions` without changing the normal user/group/other permissions? IIRC, `compinit` won't mind that.

> For clarity, let me re-state that my .zshrc invokes compinit with `-i`, so already should be bypassing the security checks.

Note that `compinit -i` is for "ignore"; it silently ignores all insecure directories. So you won't get warnings, but if it doesn't like your permission setup, you won't get the `brew` completion definitions either. It might still be skipping these directories, and just not issuing a warning.

If you trust your system setup, you could use `compinit -u` instead of `-i` to pick up the definitions regardless of permissions. This is probably what I'd do myself.
